File: src/services/api/insideSales.ts
import { apiRequest } from './client';

// ── Types ──────────────────────────────────────────────────────────────────

export interface LeadPlanPurchase {
  planName: string | null;
  durationOrQuantity: string | null;
  programPrice: string | null;
  invoiceAmount: string | null;
  paymentId: string | null;
  paymentDateAndTime: string | null;
  planAssignedAt: string | null;
  signUpDate: string | null;
  programStartDate: string | null;
  programEndDate: string | null;
  planIncludesCgm: string | null;
  cgm: string | null;
  cgmBrand: string | null;
  sensorCount: string | null;
  transmitterCount: string | null;
  bcaDevice: string | null;
  nutraceuticalsSold: string | null;
  salesTeam: string | null;
  deviceAwbNumber: string | null;
  leadConversionDate: string | null;
}

/**
 * One `dim_lead` row in the manifest `{structural columns + attributes
 * JSONB}` shape (Phase 11E). Identity + current-state are typed structural
 * columns; `attributesAtFirstSeen` is the frozen lead-profile snapshot
 * (age_group, condition, hba1c_band, intent_to_pay, source_campaign, ...)
 * and `attributes` is the mutable current-state bag (plan_name, ...). Both
 * bags are rendered generically against `useCrmSchema` — no field is
 * flattened into a bespoke named property. PII columns
 * (`firstName/lastName/phone/email/city`) arrive pre-masked from the
 * role-aware serializer.
 */
export interface LeadListRecord {
  leadId: string;
  firstName: string | null;
  lastName: string | null;
  phone: string | null;
  email: string | null;
  city: string | null;
  prospectStage: string | null;
  repName: string | null;
  source: string | null;
  createdOn: string;
  mqlScore: number | null;
  mqlSignals: Record<string, boolean>;
  attributesAtFirstSeen: Record<string, unknown>;
  attributes: Record<string, unknown>;
}

export interface LeadListResponse {
  leads: LeadListRecord[];
  total: number;
  page: number;
  pageSize: number;
  freshness: CollectionFreshness;
}

/**
 * One `fact_lead_activity` (call) row in the manifest `{structural columns
 * + attributes JSONB}` shape (Phase 11E). Typed structural columns at the
 * top level; the call-specific payload (direction, status,
 * duration_seconds, recording_url, phone_number, display_number,
 * call_notes, call_session_id, rep_email, has_recording, event_code) lives
 * in `attributes` and is rendered generically against `useCrmSchema`. PII
 * keys inside `attributes` arrive pre-masked from the role-aware
 * serializer.
 */
export interface CallRecord {
  activityId: string;
  leadId: string;
  repName: string | null;
  eventCode: number | null;
  activityType: string;
  callStartTime: string;
  createdOn: string;
  attributes: Record<string, unknown>;
  lastEvalScore?: number;
  evalCount?: number;
}

export interface CallListResponse {
  calls: CallRecord[];
  total: number;
  page: number;
  pageSize: number;
  freshness: CollectionFreshness;
}

export interface CollectionFreshness {
  lastSyncedAt: string | null;
  syncInProgress: boolean;
  stale: boolean;
}

export interface CallFilters {
  agents: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  leadId: string[];
  direction: string;
  status: string;
  hasRecording: boolean;
  eventCodes: string;
  durationMin: string;
  durationMax: string;
}

export interface LeadCallRecord {
  activityId: string;
  callTime: string;
  repName: string | null;
  durationSeconds: number;
  status: string;
  recordingUrl: string | null;
  evalScore: number | null;
  isCounseling: boolean;
}

export interface LeadEvalHistoryEntry {
  id: string;
  threadId: string;
  runId: string;
  result: Record<string, unknown>;
  createdAt: string;
}

export interface LeadDetailFullResponse {
  leadId: string;
  firstName: string;
  lastName: string | null;
  phone: string;
  email: string | null;
  prospectStage: string;
  city: string | null;
  ageGroup: string | null;
  condition: string | null;
  hba1cBand: string | null;
  bloodSugarBand: string | null;
  diabetesDuration: string | null;
  currentManagement: string | null;
  goal: string | null;
  intentToPay: string | null;
  jobTitle: string | null;
  preferredCallTime: string | null;
  repName: string | null;
  source: string | null;
  sourceCampaign: string | null;
  createdOn: string;
  mqlScore: number;
  mqlSignals: Record<string, boolean>;
  frtSeconds: number | null;
  totalDials: number;
  connectRate: number | null;
  counselingCount: number;
  counselingRate: number | null;
  callbackAdherenceSeconds: number | null;
  leadAgeDays: number;
  daysSinceLastContact: number | null;
  callHistory: LeadCallRecord[];
  historyTruncated: boolean;
  evalHistory: LeadEvalHistoryEntry[];
  plan: LeadPlanPurchase;
}

// ── API functions ──────────────────────────────────────────────────────────

export interface LeadFilters {
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  agents: string[];
  stage: string[];
  mqlMin: string;
  condition: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  city: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  leadId: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  phone: string[];
  /** Multi-select via the suggestions endpoint; CSV-joined on the wire. */
  planName: string[];
  q: string;
}

export type CallQueryScope = 'page' | 'all';
export type InsideSalesCollectionFamily = 'calls' | 'leads';

export interface CollectionSyncStatus {
  lastSuccessAt: string | null;
  lastAttemptAt: string | null;
  lastStatus: 'running' | 'completed' | 'failed' | 'cancelled' | null;
  lastError: string | null;
  syncInProgress: boolean;
}

export async function fetchCollectionStatus(
  sourceFamily: InsideSalesCollectionFamily,
): Promise<CollectionSyncStatus> {
  return apiRequest<CollectionSyncStatus>(
    `/api/inside-sales/collections/${encodeURIComponent(sourceFamily)}/status`,
  );
}

export type SuggestionField =
  | 'lead_id'
  | 'phone'
  | 'rep_name'
  | 'city'
  | 'stage'
  | 'plan_name';

export async function fetchCollectionSuggestions(
  sourceFamily: InsideSalesCollectionFamily,
  field: SuggestionField,
  q: string,
  limit = 20,
): Promise<string[]> {
  const params = new URLSearchParams({ field, limit: String(limit) });
  const trimmed = (q ?? '').trim();
  if (trimmed) params.set('q', trimmed);
  const res = await apiRequest<{ values: string[] }>(
    `/api/inside-sales/collections/${encodeURIComponent(sourceFamily)}/suggestions?${params.toString()}`,
  );
  return res.values ?? [];
}

function buildCallSearchParams(
  filters: CallFilters,
  page: number,
  pageSize: number,
  scope: CallQueryScope,
): URLSearchParams {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });

  if (scope !== 'page') {
    params.set('scope', scope);
  }
  if (filters.agents && filters.agents.length > 0) params.set('agents', filters.agents.join(','));
  if (filters.leadId && filters.leadId.length > 0) params.set('lead_id', filters.leadId.join(','));
  if (filters.direction) params.set('direction', filters.direction);
  if (filters.status) params.set('status', filters.status);
  if (filters.hasRecording) params.set('has_recording', 'true');
  if (filters.durationMin) params.set('duration_min', filters.durationMin);
  if (filters.durationMax) params.set('duration_max', filters.durationMax);
  if (filters.eventCodes) params.set('event_codes', filters.eventCodes);

  return params;
}

export async function fetchCalls(
  filters: CallFilters,
  page: number,
  pageSize: number,
  options?: { scope?: CallQueryScope },
): Promise<CallListResponse> {
  const scope = options?.scope ?? 'page';
  const params = buildCallSearchParams(filters, page, pageSize, scope);
  return apiRequest<CallListResponse>(`/api/inside-sales/calls?${params.toString()}`);
}

export async function fetchCallsForSelection(
  filters: CallFilters,
  pageSize = 500,
): Promise<CallListResponse> {
  return fetchCalls(filters, 1, pageSize, { scope: 'all' });
}

export async function fetchLeads(
  filters: LeadFilters,
  page: number,
  pageSize: number,
): Promise<LeadListResponse> {
  const params = new URLSearchParams({
    page: String(page),
    page_size: String(pageSize),
  });
  const q = (filters.q ?? '').trim();
  if (filters.agents && filters.agents.length > 0) params.set('agents', filters.agents.join(','));
  if (filters.stage && filters.stage.length > 0) params.set('stage', filters.stage.join(','));
  if (filters.mqlMin) params.set('mql_min', filters.mqlMin);
  if (filters.condition && filters.condition.length > 0) params.set('condition', filters.condition.join(','));
  if (filters.city && filters.city.length > 0) params.set('city', filters.city.join(','));
  if (filters.leadId && filters.leadId.length > 0) params.set('lead_id', filters.leadId.join(','));
  if (filters.phone && filters.phone.length > 0) params.set('phone', filters.phone.join(','));
  if (filters.planName && filters.planName.length > 0) params.set('plan_name', filters.planName.join(','));
  if (q) params.set('q', q);

  return apiRequest<LeadListResponse>(`/api/inside-sales/leads?${params.toString()}`);
}

export async function fetchLeadDetail(
  leadId: string,
): Promise<LeadDetailFullResponse> {
  return apiRequest<LeadDetailFullResponse>(`/api/inside-sales/leads/${leadId}/detail`);
}

File: backend/app/schemas/inside_sales.py
"""Schemas for Inside Sales API."""

from datetime import datetime
from typing import Any, Optional
from pydantic import Field
from app.schemas.base import CamelModel


class CallRecord(CamelModel):
    """One ``fact_lead_activity`` (call) row in the manifest
    ``{structural columns + attributes JSONB}`` shape (Phase 11E).

    Typed structural columns at the top level; the call-specific payload
    (direction, status, duration_seconds, recording_url, phone_number,
    display_number, call_notes, call_session_id, rep_email, has_recording,
    event_code) lives in ``attributes`` and is rendered generically by the
    frontend against ``useCrmSchema``. PII keys inside ``attributes`` are
    masked by the role-aware serializer."""
    activity_id: str
    lead_id: str
    rep_name: Optional[str] = None
    event_code: Optional[int] = None
    activity_type: str
    call_start_time: str
    created_on: str
    attributes: dict[str, Any] = Field(default_factory=dict)
    last_eval_score: Optional[float] = None
    eval_count: int = 0


class CollectionFreshness(CamelModel):
    last_synced_at: datetime | None = None
    sync_in_progress: bool = False
    stale: bool = True


class CallListResponse(CamelModel):
    calls: list[CallRecord]
    total: int
    page: int
    page_size: int
    freshness: CollectionFreshness = Field(default_factory=CollectionFreshness)


class LeadDetailResponse(CamelModel):
    lead_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None


class LeadPlanPurchase(CamelModel):
    """Plan-purchase surface derived from LSQ ``mx_*`` fields.

    All optional — a lead in a pre-converted stage will have these null.
    Values are stored on the ``analytics.crm_lead_record.raw_payload`` JSONB
    blob and extracted at API-response time.
    """
    plan_name: Optional[str] = None
    duration_or_quantity: Optional[str] = None
    program_price: Optional[str] = None
    invoice_amount: Optional[str] = None
    payment_id: Optional[str] = None
    payment_date_and_time: Optional[str] = None
    plan_assigned_at: Optional[str] = None
    sign_up_date: Optional[str] = None
    program_start_date: Optional[str] = None
    program_end_date: Optional[str] = None
    plan_includes_cgm: Optional[str] = None
    cgm: Optional[str] = None
    cgm_brand: Optional[str] = None
    sensor_count: Optional[str] = None
    transmitter_count: Optional[str] = None
    bca_device: Optional[str] = None
    nutraceuticals_sold: Optional[str] = None
    sales_team: Optional[str] = None
    device_awb_number: Optional[str] = None
    lead_conversion_date: Optional[str] = None


class LeadListRecord(CamelModel):
    """One ``dim_lead`` row in the manifest ``{structural columns +
    attributes JSONB}`` shape (Phase 11E).

    Identity + current-state are typed structural columns (the
    ``firstName/lastName/phone/email/city`` PII columns are masked by the
    role-aware serializer). ``attributesAtFirstSeen`` is the frozen
    lead-profile snapshot (age_group, condition, hba1c_band, intent_to_pay,
    source_campaign, ...); ``attributes`` is the mutable current-state bag
    (plan_name, ...). MQL is assembled from ``fact_lead_signal``. The
    frontend renders the bags generically against ``useCrmSchema``."""
    lead_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    city: Optional[str] = None
    prospect_stage: Optional[str] = None
    rep_name: Optional[str] = None
    source: Optional[str] = None
    created_on: str
    mql_score: Optional[int] = None
    mql_signals: dict[str, bool] = Field(default_factory=dict)
    attributes_at_first_seen: dict[str, Any] = Field(default_factory=dict)
    attributes: dict[str, Any] = Field(default_factory=dict)


class LeadListResponse(CamelModel):
    leads: list[LeadListRecord]
    total: int
    page: int
    page_size: int
    freshness: CollectionFreshness = Field(default_factory=CollectionFreshness)


class CollectionRefreshRequest(CamelModel):
    # ``sync_mode`` lets ops trigger either a one-shot 90-day bootstrap
    # (``date_range`` with explicit ``dateFrom``/``dateTo``) or an
    # on-demand delta (``incremental``). ``bootstrap`` is sugar for a
    # 90-day ``date_range`` with the plan's defaults.
    sync_mode: str | None = None
    date_from: str | None = None
    date_to: str | None = None
    event_codes: str | None = None
    overlap_minutes: int | None = None


class CollectionRefreshResponse(CamelModel):
    job_id: str
    source_family: str
    sync_mode: str
    status: str


class CollectionRunEntry(CamelModel):
    """One row in the last-N ``analytics.log_crm_source_sync`` list surfaced to ops."""
    id: str
    sync_mode: str
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
    watermark_from: str | None = None
    watermark_to: str | None = None
    records_scanned: int = 0
    records_upserted: int = 0
    records_failed: int = 0
    is_scheduled_run: bool = False
    error_message: str | None = None


class CollectionRunsResponse(CamelModel):
    source_family: str
    runs: list[CollectionRunEntry]


class CollectionSyncStatus(CamelModel):
    """Durable freshness signal for a collection. Read from ``analytics.log_crm_source_sync``.

    ``lastSuccessAt`` is the most recent ``completed`` sync. ``lastAttemptAt``
    is the most recent attempt regardless of outcome. ``lastStatus`` /
    ``lastError`` describe that attempt so the UI can render failure state
    after a page reload (frontend cache is not durable across reloads).
    ``syncInProgress`` is true when any sync is currently ``running``.
    """
    last_success_at: datetime | None = None
    last_attempt_at: datetime | None = None
    last_status: str | None = None  # 'running' | 'completed' | 'failed' | 'cancelled'
    last_error: str | None = None
    sync_in_progress: bool = False


class LeadEvalHistoryEntry(CamelModel):
    """One evaluation record for a lead's call history."""
    id: str
    thread_id: str
    run_id: str
    result: dict                    # raw evaluator JSON
    created_at: str


class LeadCallRecord(CamelModel):
    activity_id: str
    call_time: str
    rep_name: Optional[str] = None
    duration_seconds: int = 0
    status: str
    recording_url: Optional[str] = None
    eval_score: Optional[float] = None
    is_counseling: bool = False              # duration_seconds >= 600


class LeadDetailFullResponse(CamelModel):
    # Profile
    lead_id: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    prospect_stage: str
    city: Optional[str] = None
    age_group: Optional[str] = None
    condition: Optional[str] = None
    hba1c_band: Optional[str] = None
    blood_sugar_band: Optional[str] = None
    diabetes_duration: Optional[str] = None
    current_management: Optional[str] = None
    goal: Optional[str] = None
    intent_to_pay: Optional[str] = None
    job_title: Optional[str] = None
    preferred_call_time: Optional[str] = None
    rep_name: Optional[str] = None
    source: Optional[str] = None
    source_campaign: Optional[str] = None
    created_on: str
    # MQL
    mql_score: int = 0
    mql_signals: dict[str, bool] = Field(default_factory=dict)
    # Computed metrics
    frt_seconds: Optional[int] = None
    total_dials: int = 0
    connect_rate: Optional[float] = None
    counseling_count: int = 0
    counseling_rate: Optional[float] = None
    callback_adherence_seconds: Optional[int] = None
    lead_age_days: int = 0
    days_since_last_contact: Optional[int] = None
    # Call history
    call_history: list[LeadCallRecord] = Field(default_factory=list)
    history_truncated: bool = False
    # Eval history
    eval_history: list[LeadEvalHistoryEntry] = Field(default_factory=list)
    # Plan-purchase surface. Null fields for pre-converted leads.
    plan: LeadPlanPurchase = Field(default_factory=LeadPlanPurchase)
